ksonnet-gen/ksonnet/emit.go

```diff
@@ -508,18 +519,83 @@ func (ao *apiObject) emitAsRefMixins(
 	m.writeLine("},")
 }
 
-func (ao *apiObject) emitConstructor(m *indentWriter) {
-	if dm, ok := ao.properties[constructorName]; ok {
+func (ao *apiObject) emitConstructors(m *indentWriter) {
+	k8sVersion := ao.root().spec.Info.Version
+	path := ao.parsedName.Unparse()
+
+	specs, ok := kubeversion.ConstructorSpec(k8sVersion, path)
+	if !ok {
+		ao.emitConstructor(m, constructorName, []kubeversion.CustomConstructorParam{})
+		return
+	}
+
+	for _, spec := range specs {
+		ao.emitConstructor(m, spec.ID, spec.Params)
+	}
+}
+
+func (ao *apiObject) emitConstructor(
+	m *indentWriter, id string, params []kubeversion.CustomConstructorParam,
+) {
+	// Panic if a function with the constructor's name already exists.
+	specName := kubespec.PropertyName(id)
+	if dm, ok := ao.properties[specName]; ok {
 		log.Panicf(
-			"Attempted to create constructor, but 'new' property already existed at '%s'",
-			dm.path)
+			"Attempted to create constructor, but '%s' property already existed at '%s'",
+			specName, dm.path)
 	}
 
+	// Default body of the constructor. Usually either `apiVersion +
+	// kind` or `{}`.
+	var defaultSetters []string
 	if ao.isTopLevel {
-		m.writeLine(fmt.Sprintf("%s():: apiVersion + kind,", constructorName))
+		defaultSetters = specialPropertiesList
 	} else {
-		m.writeLine(fmt.Sprintf("%s():: {},", constructorName))
+		defaultSetters = []string{"{}"}
 	}
+
+	// Build parameters and body of constructor. Considering the example
+	// of the constructor of `v1.Container`:
+	//
+	//   new(name, image):: self.name(name) + self.image(image),
+	//
+	// Here we want to (1) assemble the parameter list (i.e., `name` and
+	// `image`), as well as the body (i.e., the calls to `self.name` and
+	// so on).
+	paramLiterals := []string{}
+	setters := defaultSetters
+	for _, param := range params {
+		// Add the param to the param list, including default value if
+		// applicable.
+		if param.DefaultValue != nil {
+			paramLiterals = append(
+				paramLiterals, fmt.Sprintf("%s=%s", param.ID, *param.DefaultValue))
+		} else {
+			paramLiterals = append(paramLiterals, param.ID)
+		}
+
+		// Add an element to the body (e.g., `self.name` above).
+		if param.RelativePath == nil {
+			prop, ok := ao.properties[kubespec.PropertyName(param.ID)]
+			if !ok {
+				log.Panicf(
+					"Attempted to create constructor, but property '%s' does not exist",
+					param.ID)
+			}
+			setters = append(
+				setters, fmt.Sprintf("self.%s(%s)", prop.name, param.ID))
+		} else {
+			// TODO(hausdorff): We may want to verify this relative path
+			// exists.
+			setters = append(
+				setters, fmt.Sprintf("self.%s(%s)", *param.RelativePath, param.ID))
+		}
+	}
+
+	// Write out constructor.
+	paramsText := strings.Join(paramLiterals, ", ")
+	bodyText := strings.Join(setters, " + ")
+	m.writeLine(fmt.Sprintf("%s(%s):: %s,", specName, paramsText, bodyText))
```

Comment:
Go brought this up in [PR #22](https://github.com/ksonnet/ksonnet-lib/pull/22#discussion_r119999387) as well. I'm not saying it was the right decision necessarily, but the reasony of why we didn't do it this way is because I don't want to check `err` after every call to `writeLine`, and I don't see any major disadvantages to not using `io.Writer`.

So, what do you think? Do you think it's worth rearchitecting the code to make this an `io.Writer`?